### PR TITLE
FileEditor: add a 'source' option in constructor

### PR DESCRIPTION
**Pull request by jouvin**
This PR adds to the FileEditor the capability to open a reference version of the file instead of the file itself, if the reference file happens to be newer than the file. The typical use case is a file provided by a RPM that you edit with Quattor. You may want that if a new version of the RPM installs a new file as `.rpm-new`, your next run of Quattor applies modifications to this new version (this is particularly useful with the RuleBasedEditor). 

The idea is that you do it by adding a 'source' option whose value is a file name. If the source doesn't exist, it is just ignored. In this case the file will be used if it exists (current behaviour).

### CODE REVIEW

**Comment by jouvin** [comment]
In the current version of the PR, the associated unit test `test-caffileeditor_source.t` doesn't work because `LC::File::file_contents` returns an `undef` value instead of the contents. I have no clue why this happens in the context of the test. Any help by the Perl experts appreciated!

File: src/main/perl/FileEditor.pm [hunk 1]
```diff
@@ -58,13 +58,38 @@ sub _is_valid_source
...
+{
+    my ($self) = @_;
+    my $is_newer = 0;   # Assume false
+    if ( $self->_is_valid_source(*$self->{options}->{source}) ) {
+        if ( !$self->_is_valid_source(*$self->{filename}) || 
+             ((stat(*$self->{options}->{source}))[9] > (stat(*$self->{filename}))[9]) ) {
+          $is_newer = 1
+        }
+    }
+    if ( $is_newer ) {
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: too much duplicated code, make this one line
```perl
*$self->{LOG}->debug(2, "Source file (", *$self->{options}->{source}, ") is ",($is_newer? "newer" : "older")," than ", *$self->{filename}, ": ",($is_newer ? "use" : "ignore")," it");
```

File: src/main/perl/FileEditor.pm [hunk 2]
```diff
@@ -58,13 +58,38 @@ sub _is_valid_source
...
 sub new
 {
     my $class = shift;
     my $self = $class->SUPER::new (@_);
-    if ($self->_is_valid_source(*$self->{filename})) {
-        my $txt = LC::File::file_contents (*$self->{filename});
+    my $txt;
+    my ($path, %opts) = @_;
+
+    *$self->{options}->{source} = $opts{source} if exists ($opts{source});
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: needs an entry in the pod as well

File: src/main/perl/FileEditor.pm [hunk 3]
```diff
@@ -58,13 +58,38 @@ sub _is_valid_source
     return -f $fn;
 }
 
+sub _is_reference_newer
+{
+    my ($self) = @_;
+    my $is_newer = 0;   # Assume false
+    if ( $self->_is_valid_source(*$self->{options}->{source}) ) {
+        if ( !$self->_is_valid_source(*$self->{filename}) || 
+             ((stat(*$self->{options}->{source}))[9] > (stat(*$self->{filename}))[9]) ) {
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: i'm not sure about `.rpm-new`, but rpms maintain timestamps of files, so files added by rpms at time X might have older timestamps.
also CAF (via LC) allows to set the mtime, so this is might give some unwanted sideeffects.
(but i like the idea of the PR)

File: src/main/perl/FileEditor.pm [hunk 4]
```diff
@@ -58,13 +58,38 @@ sub _is_valid_source
...
 {
     my $class = shift;
     my $self = $class->SUPER::new (@_);
-    if ($self->_is_valid_source(*$self->{filename})) {
-        my $txt = LC::File::file_contents (*$self->{filename});
+    my $txt;
+    my ($path, %opts) = @_;
+
+    *$self->{options}->{source} = $opts{source} if exists ($opts{source});
+    if ( exists(*$self->{options}->{source}) && $self->_is_reference_newer() ) {
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: the `exists()` is covered by the `_is_reference_newer`

File: src/main/perl/FileEditor.pm (hunk 1 above)
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: also needs `... if *$self->{LOG};` or the conditional logger from #154

File: src/main/perl/FileEditor.pm [hunk 5]
```diff
@@ -58,13 +58,38 @@ sub _is_valid_source
...
+        $txt = LC::File::file_contents (*$self->{options}->{source});
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: this needs to be logged at least with `verbose` that source is used

Comment (stdweird) [inline src/main/perl/FileEditor.pm]: also avoid code duplication. best to only modify the filename to read from, and use the `LC::File::file_contents`
even better, let `_is_reference_newer` return the correct filename to read from

File: src/main/perl/FileEditor.pm [hunk 6]
```diff
@@ -58,13 +58,38 @@ sub _is_valid_source
...
+    } elsif ($self->_is_valid_source(*$self->{filename})) {
+        $txt = LC::File::file_contents (*$self->{filename});
+    }
+    if ( $txt ) {
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: has to be `if (defined($txt))`

File: src/test/perl/test-caffileeditor_source.t [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+use lib "$Bin/modules";
+use testapp;
+use CAF::FileEditor;
+use Test::More tests => 3;
+use Test::Quattor::Object;
+use Carp qw(confess);
+use File::Path;
+use File::Temp qw(tempfile);
+
+my $testdir = 'target/test/editor';
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: the testdir has to be unique for the test, we cannot risk using directories with files from previous tests

File: src/test/perl/test-caffileeditor_source.t [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+use constant TEXT => <<EOF;
+En un lugar de La Mancha, de cuyo nombre no quiero acordarme
+no ha tiempo que vivía un hidalgo de los de lanza en astillero...
+EOF
+use constant HEADTEXT => <<EOF;
+... adarga antigua, rocín flaco y galgo corredor.
+EOF
+
+my $fh;
+my $this_app = testapp->new ($0, qw (--verbose));
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: remove the testapp code, you are not using it

File: src/test/perl/test-caffileeditor_source.t [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+*testapp::error = sub {
+    my $self = shift;
+    $self->{ERROR} = @_;
+};
+
+# Create a file and check that it is empty
+($fh, $filename) = tempfile(DIR => $testdir);
+$fh->close();
+$fh = CAF::FileEditor->new($filename, log => $obj);
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: use `CAF::FileReader` to read the contents of a file, this is confusing

File: src/test/perl/test-caffileeditor_source.t [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+my $testdir = 'target/test/editor';
+mkpath($testdir);
+(undef, my $filename) = tempfile(DIR => $testdir);
+
+use constant TEXT => <<EOF;
+En un lugar de La Mancha, de cuyo nombre no quiero acordarme
+no ha tiempo que vivía un hidalgo de los de lanza en astillero...
+EOF
+use constant HEADTEXT => <<EOF;
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: remove if not used

File: src/test/perl/test-caffileeditor_source.t [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+is("$fh","","Existing file ($filename) empty");
+$fh->close();
+
+# Check that reference file contents is used as the initial contents when it
+# is newer than the file edited.
+sleep 5;
+my ($ref_fh, $ref_filename) = tempfile(DIR => $testdir);
+print $ref_fh TEXT;
+$ref_fh->close();
+$fh = CAF::FileEditor->new($filename, log => $obj, source => $ref_filename);
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: use FileReader (looks odd, but is a subclass of Editor); but why do you cancel?

File: src/test/perl/test-caffileeditor_source.t [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+($fh, $filename) = tempfile(DIR => $testdir);
+$fh->close();
+$fh = CAF::FileEditor->new($filename, log => $obj);
+$fh->cancel;
+is("$fh","","Existing file ($filename) empty");
+$fh->close();
+
+# Check that reference file contents is used as the initial contents when it
+# is newer than the file edited.
+sleep 5;
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: set the mtime of file instead of sleeping

File: src/test/perl/test-caffileeditor_source.t [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+print $ref_fh TEXT;
+$ref_fh->close();
+$fh = CAF::FileEditor->new($filename, log => $obj, source => $ref_filename);
+$fh->cancel;
+is("$fh",TEXT,"Reference file ($filename) contents used");
+$fh->close();
+
+# Check that reference file contents is not used as the initial contents when it
+# is older than the file edited.
+sleep 5;
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: change the original file mtime, it's easier

File: src/test/perl/test-caffileeditor_source.t [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+my ($new_fh, $new_filename) = tempfile(DIR => $testdir);
+$new_fh->close();
+$new_fh = CAF::FileEditor->new($new_filename, log => $obj, source => $ref_filename);
+$new_fh->cancel;
+is("$new_fh","","Existing file ($new_filename) contents used");
+$fh->close();
+
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: test is missing
* existing of source attribute
* test of private method

File: src/main/perl/FileEditor.pm (hunk 3 above)
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: can you also add a comment what the 10th entry is in the stat result?

Comment (jouvin) [inline src/main/perl/FileEditor.pm]: 10th enty is the mod time and should not be affected by the creation time in the RPM.

File: src/test/perl/test-caffileeditor_source.t (hunk 11 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: No reason in fact!

File: src/test/perl/test-caffileeditor_source.t (hunk 12 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: The pb is that stat() as a 1s resolution... I plan to reduce it to s or 3s.

**Comment by jouvin** [comment]
@stdweird thanks for all your comments, I'll implement them (as usual!). I was aware that the code was not polished. I concentrated on getting the test working and unfortunately I don't see anything in your comment that may help... Even the `defined($text)` is not the cause of the problem as `$txt` should be non empty in the context of the failing text. In fact is you print `$txt` after calling `LC::File::file_contents` (using `debug()` or friends), you will get a warning that `$txt` is undefined. No clue why as this is exactly the existing code that is working (but probably not really tested as it is generally mocked).

File: src/test/perl/test-caffileeditor_source.t (hunk 7 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: I don't think it is a problem as the file name is generated (based on a time stamp I guess). And it makes the debugging easier IMO.

File: src/test/perl/test-caffileeditor_source.t (hunk 9 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: I prefer to really test `FileEditor` as this is a test for it.

File: src/test/perl/test-caffileeditor_source.t (hunk 11 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: In fact it is needed, at least the first one, as the file is reused in next step and we want it to be empty (else the test will succeed even if the code fails).

File: src/test/perl/test-caffileeditor_source.t (hunk 12 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: I don't think that manipulating mtime is easier/cleaner.

File: src/test/perl/test-caffileeditor_source.t (hunk 13 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: I tend to disagree as I first need to retrieve the existing one...

File: src/main/perl/FileEditor.pm (hunk 1 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: Honnestly I don't buy this suggestion as a simplification...

**Comment by jouvin** [comment]
@stdweird I did more tests to understand the problem of `LC::File::file_contents`. I have been able to reproduce the problem with the original `FileEditor` code (6b587a4e7b19234d1924b09fe30d6543e40245df). If you execute `test-caffileeditor_source.t` (after removing the 4 lines related to testing `source` and `_is_reference_newer()`) you obtain the same failing tests which is unexpected for the last test (file newer than reference). And if you add in the constructor after the call to `LC::File::file_contents` the following line:
```
        *$self->{LOG}->debug(2,*$self->{filename}." contents: $txt");
```

You get a warning that `$txt`  is uninitialized. As for me, the problem is related to the execution of `LC::File::file_contents` in the unit test context as this piece of code works well in the real life! No clue why...

**Comment by jouvin** [comment]
Workaround for the failing `LC::File::file_contents` in the test context: mock it... Not glorious... but I didn't find a better solution as clearly (see previous comment) the problem is `LC::File::file_contents` itself, not the calling code.

File: src/main/perl/FileEditor.pm [hunk 15]
```diff
@@ -58,13 +66,45 @@ sub _is_valid_source
     return -f $fn;
 }
 
+sub _is_reference_newer
```
Comment (ned21) [inline src/main/perl/FileEditor.pm]: This sub routine lacks any POD. It also hardcodes references to the internal data structures which makes it hard to follow: I have to know what *$self->{options}->{source} and *$self->{filename} refer to in order to grok it. Could it be made more general, so that the things/references to be compared are passed as arguments? Then the function could conceivably be reused in other places too.

Comment (jouvin) [inline src/main/perl/FileEditor.pm]: This is the usual stuff in CAF::Filexxx. There are already quite a lot of it around. I can add a comment that you should look at `CAF::FileWriter` for details. I can add a POD but this is a very internal method and as there were none for `_is_valid` I decided to do the same...

File: src/main/perl/FileEditor.pm [hunk 16]
```diff
@@ -47,7 +47,15 @@ the class constructor.
 =item new
 
 Returns a new object it accepts the same arguments as the constructor
-for C<CAF::FileWriter>
+for C<CAF::FileWriter> with one additional option: C<source>. This
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: make this extendable
```perl
with following addiotional options:
=over

=item source: ...

=back
```

File: src/main/perl/FileEditor.pm [hunk 17]
```diff
@@ -47,7 +47,15 @@ the class constructor.
...
+option, when present, must be a file name whose contents will be used
+as the initial contents for the edited file if the source modification time
+is more recent than the edited file modification time. This allows to rebuild
+the file contents based on a new version of the reference/template file.
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: remove the `/template`, nothing is templated here. i would also not use the term `reference` but stick with `source` to avoid confusion

File: src/main/perl/FileEditor.pm [hunk 18]
```diff
@@ -47,7 +47,15 @@ the class constructor.
...
 
 Returns a new object it accepts the same arguments as the constructor
-for C<CAF::FileWriter>
+for C<CAF::FileWriter> with one additional option: C<source>. This
+option, when present, must be a file name whose contents will be used
+as the initial contents for the edited file if the source modification time
+is more recent than the edited file modification time. This allows to rebuild
+the file contents based on a new version of the reference/template file. 
+
+An example is:
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: i'm not sure what the example adds here. it's an option like any other one

File: src/main/perl/FileEditor.pm [hunk 19]
```diff
@@ -58,13 +66,45 @@ sub _is_valid_source
     return -f $fn;
 }
 
+sub _is_reference_newer
+{
+    my ($self) = @_;
+    my $is_newer = 0;   # Assume false
+    if (  exists(*$self->{options}->{source}) && $self->_is_valid_source(*$self->{options}->{source}) ) {
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: ok, this is potentially dangerous, as `FileReader` inherits from Editor, but `valid_source` for reader can be a pipe, and the stat code below will not like that. so this PR needs modifcations for FileReader (unless you want editor to support pipes as `source`, which could make sense).

File: src/main/perl/FileEditor.pm (hunk 15 above)
Comment (ned21) [inline src/main/perl/FileEditor.pm]: That was really my point - this is too much of an internal/single use specific function. If it was made more general and independent of the internal structures of this class then it could potentially be moved else where and used by other parts of the code base. It would also be easier to understand and test since the reader/test would only need to understand the function and not the surrounding state.

File: src/main/perl/FileEditor.pm [hunk 20]
```diff
@@ -58,13 +66,45 @@ sub _is_valid_source
...
+        if ( !$self->_is_valid_source(*$self->{filename}) || 
+             ((stat(*$self->{options}->{source}))[9] > (stat(*$self->{filename}))[9]) ) {
+          $is_newer = 1
+        }
+    }
+    #FIXME: replace by $self->debug() after PR #154 has been merged...
+    if ( *$self->{LOG} ) {
+        if ( $is_newer ) {
+            # As this is a non reproducible event, be sure to log it when it happens
+            *$self->{LOG}->info("File ", *$self->{filename}, " contents reset to reference file (", *$self->{options}->{source}, ") contents.");
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: both messages are not correct, as this method returns a simple boolean, it does not do anything (like resetting or ignoring).
i would just report the value of `is_newer` with debug 1 and report with info (nice catch on non-reproducible) later when the source is set.
ideally, the is_newer is a string with the reason why, something that can be added to the message, in this case eg. a timedifference or mtime of both files.

File: src/main/perl/FileEditor.pm [hunk 21]
```diff
@@ -58,13 +66,45 @@ sub _is_valid_source
...
 {
     my $class = shift;
     my $self = $class->SUPER::new (@_);
-    if ($self->_is_valid_source(*$self->{filename})) {
-        my $txt = LC::File::file_contents (*$self->{filename});
+    my $src_file;
+    my ($path, %opts) = @_;
+
+    *$self->{options}->{source} = $opts{source} if exists ($opts{source});
+    if ( $self->_is_reference_newer() ) {
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: the name of the method is and option are also a bity confusion: the source option is dealt with via the reference method, but the is also a `is_valid_source` method that has nothing to do with `source` option.

File: src/main/perl/FileEditor.pm [hunk 22]
```diff
@@ -58,13 +66,45 @@ sub _is_valid_source
...
+        $src_file = *$self->{options}->{source};
+    } elsif ($self->_is_valid_source(*$self->{filename})) {
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: this is differencet from the old code, where `filename` was always verified to be a valid filename, in this case, this check is skipped if source is used. if source is defined, filename should either no exist or be a file.
`CAF::Check` has methods for that, but if you don't want to use caf::Check, but the existence and directory tests in separate methods, as `-e` and `-d` can't be mocked.

File: src/test/perl/test-caffileeditor_source.t [hunk 23]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/perl
+use strict;
+use warnings;
+use FindBin qw($Bin);
+use lib "$Bin/modules";
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: this is what loads dummy lc modules, remove it and you should be able to use `LC::File::file_contents` without the mocking

File: src/test/perl/test-caffileeditor_source.t [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+    my $fname = shift;
+    my $fh;
+    open($fh, "<", $fname) || die ("failed to open $fname");
+    my $contents = join('', <$fh>);
+    return $contents;
+}
+$lcfile->mock("file_contents", \&read_file_contents);
+
+
+my $testdir = 'target/test/editor';
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: unique test dirs per test, e.g. rename this to `editor_source`

File: src/test/perl/test-caffileeditor_source.t [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+# Create a file and check that it is empty
+($fh, $filename) = tempfile(DIR => $testdir);
+$fh->close();
+$fh = CAF::FileEditor->new($filename, log => $obj);
+is("$fh","","Existing file ($filename) empty");
+$fh->close();
+
+# Check that reference file contents is used as the initial contents when it
+# is newer than the file edited.
+sleep 2;
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: set the mtime of reference/source or filename using `LC::Check::status($fn, mtime => 123456)` instead of sleeping

File: src/test/perl/test-caffileeditor_source.t [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+$ref_fh->close();
+$fh = CAF::FileEditor->new($filename, log => $obj, source => $ref_filename);
+is(*$fh->{options}->{source},$ref_filename,"File source is correctly defined");
+ok($fh->_is_reference_newer(),"Source file ($ref_filename) is newer than actual file ($filename)");
+is("$fh",$TEXT,"Reference file ($filename) contents used");
+$fh->close();
+
+# Check that reference file contents is not used as the initial contents when it
+# is older than the file edited.
+sleep 2;
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: set the mtime as above

File: src/test/perl/test-caffileeditor_source.t [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+my ($new_fh, $new_filename) = tempfile(DIR => $testdir);
+print $new_fh $ANOTHER_TEXT;
+$new_fh->close();
+$new_fh = CAF::FileEditor->new($new_filename, log => $obj, source => $ref_filename);
+is(*$new_fh->{options}->{source},$ref_filename,"File source is correctly defined");
```
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: hmm, i didn't realise this. wouldn't it be better that the source option is deleted from the options if it is not used as source? that way we have a way to determine if the source was used as source or not. with the current code, presence of the source as option is non-conclusive.

File: src/main/perl/FileEditor.pm (hunk 15 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: After thinking more about it, I prefer to keep it this way: an internal method valid only in the context of the constructor. I added comments (rather than POD) to make this explicit and explain the main attributes used.

File: src/main/perl/FileEditor.pm (hunk 17 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: `/template` removed but I think that using `source` in the text will be more confusing as one could consider that the source is the filename you passed when instantiating the `FileEditor`. and `reference` as the option name may also be very strange/confusiong... I didn't manage to find a perfect solution! I mentioned in another comment that `source` parameter value is referred internally as reference.

File: src/main/perl/FileEditor.pm (hunk 18 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: It was an attempt to help! Removed...

File: src/main/perl/FileEditor.pm (hunk 19 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: How to easily distinguish a pipe from a standard file? If we agree that supporting pipes as `source` make sense, I propose that if it is a pipe, it is always considered newer than the existing file.

File: src/main/perl/FileEditor.pm (hunk 21 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: I saw it, I'd suggest to rename `_is_valid_source()`!

File: src/main/perl/FileEditor.pm (hunk 22 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: Nice catch! Clearly `filename` must be a valid filename in any case so that the changes can be written.

File: src/test/perl/test-caffileeditor_source.t (hunk 25 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: I don't understand why... For me this is a much better test to do a real modification with a short time interval...

File: src/test/perl/test-caffileeditor_source.t (hunk 27 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: I don't have a strong opinion. But I don't think we tend to delete options and this may make the troubleshooting more difficult... This is the role of the constructor to decide what to do. If it is really important to know if the source was used, I'd prefer to add an attribute rather than remove the option....

File: src/main/perl/FileEditor.pm (hunk 22 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: @stdweird in fact I am not sure to completely understand your point. In the previous code, nothing prevented filename to be a directory for example. Just it was not read but when you were closing the file, you would have got an error. I don't think the new code is really different. But it may be good to add the test you suggest, i.e. be sure that the file wil be writable at the end. May be better to handle as a separate issue...

File: src/main/perl/FileEditor.pm (hunk 19 above)
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: uisng `-p`, see `_is_valid_source` in FileReader

File: src/main/perl/FileEditor.pm (hunk 22 above)
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: @jouvin heh, that is indeed the case. i was unaware of this. i opened https://github.com/quattor/CAF/issues/157, but i already do not see how this could be handled cleanly or what it would add :confused:

File: src/main/perl/FileEditor.pm (hunk 21 above)
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: also fine by me, do not forget to modify FileReader as well

File: src/test/perl/test-caffileeditor_source.t (hunk 25 above)
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: it adds 4 seconds to the tests, and setting the mtime by hand to some non-sensical value is much better way to control it than some random sleep value. it's as simple as
```perl
$time=time()
utime(undef, $time - 10, $filename); # make filename old enough
```

File: src/test/perl/test-caffileeditor_source.t (hunk 27 above)
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: ok, if the need ever arises, i guess we can add a source attribute to track actual source

File: src/test/perl/test-caffileeditor_source.t (hunk 23 above)
Comment (jouvin) [inline src/test/perl/test-caffileeditor_source.t]: I don't see any dummy LC modules and removing this `use lib` doesn't help if I didn't do something wrong...

File: src/main/perl/FileEditor.pm (hunk 19 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: I added pipe support for `source` but I am not sure how to unit test it... (btw, I don't think this is unit tested in FileReader too).

File: src/main/perl/FileEditor.pm (hunk 21 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: Changed to `is_valid_file` but it seems it was not detected by GitHub.

File: src/test/perl/test-caffileeditor_source.t (hunk 23 above)
Comment (stdweird) [inline src/test/perl/test-caffileeditor_source.t]: apologies again, you need code like https://github.com/quattor/CAF/blob/master/src/test/perl/kerberos-process.t#L4
the dummy LC modules are in the resources dir

File: src/main/perl/FileEditor.pm [hunk 28]
```diff
@@ -47,24 +47,79 @@ the class constructor.
...
 
 Returns a new object it accepts the same arguments as the constructor
-for C<CAF::FileWriter>
+for C<CAF::FileWriter> with one additional option: 
+
+=over
+
+=item source 
+
+This option, when present, must be a file name whose contents will be used
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: with current code, the source for Editor can also be a pipe (and maybe mention that if it is a pipe, it is always considered more recent)

File: src/main/perl/FileEditor.pm [hunk 29]
```diff
@@ -47,24 +47,79 @@ the class constructor.
...
+# internally).
+sub _is_reference_newer
+{
+    my ($self) = @_;
+    my $is_newer = 0;   # Assume false
+    if (  exists(*$self->{options}->{source}) ) {
+        # It is valid for the source value to be a pipe: in this case consider it
+        # as newer than an existing file.
+        if ( -p *$self->{options}->{source} ) {
+          $is_newer = 1
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: indentation is wrong?

File: src/main/perl/FileEditor.pm [hunk 30]
```diff
@@ -47,24 +47,79 @@ the class constructor.
...
+        } elsif ( $self->_is_valid_file(*$self->{options}->{source}) ) {
+        # stat()[9] is modification time
```
Comment (stdweird) [inline src/main/perl/FileEditor.pm]: indentation is wrong

**Comment by stdweird** [comment]
@jouvin some minor remarks left, LGTM otherwise

File: src/main/perl/FileEditor.pm (hunk 28 above)
Comment (jouvin) [inline src/main/perl/FileEditor.pm]: Done.

**Comment by jouvin** [comment]
Last comments hopefully addresssed.

**Comment by stdweird** [comment]
@jouvin thanks a lot!
